cleanrl/rnd_ppo.py

```diff
@@ -0,0 +1,1079 @@
+            ext_returns = ext_advantages + ext_values
+            int_returns = int_advantages + int_values
+        else:
+            ext_returns = torch.zeros_like(rewards).to(device)
+            int_returns = torch.zeros_like(curiosity_rewards).to(device)
+            for t in reversed(range(args.num_steps)):
+                if t == args.num_steps - 1:
+                    ext_nextnonterminal = 1.0 - next_done
+                    int_nextnonterminal = 1.0
+                    ext_next_return = last_value_ext
+                    int_next_return = last_value_int
+                else:
+                    ext_nextnonterminal = 1.0 - dones[t+1]
+                    int_nextnonterminal = 1.0
+                    ext_next_return = ext_returns[t+1]
+                    int_next_return = int_returns[t+1]
+                ext_returns[t] = rewards[t] + args.gamma * ext_nextnonterminal * ext_next_return
+                int_returns[t] = curiosity_rewards[t] + args.int_gamma * int_nextnonterminal * int_next_return
+            ext_advantages = ext_returns - ext_values
+            int_advantages = int_returns - int_values
+
+    # flatten the batch
+    b_obs = obs.reshape((-1,)+envs.observation_space.shape)
+    b_logprobs = logprobs.reshape(-1)
+    b_actions = actions.reshape((-1))
+    b_ext_advantages = ext_advantages.reshape(-1)
+    b_int_advantages = int_advantages.reshape(-1)
+    b_ext_returns = ext_returns.reshape(-1)
+    b_int_returns = int_returns.reshape(-1)
+    b_ext_values = ext_values.reshape(-1)
+
+    b_advantages = b_int_advantages * args.int_coef + b_ext_advantages * args.ext_coef
+    del b_int_advantages, b_ext_advantages
+
+    obs_rms.update(b_obs.data.cpu().numpy()[:,3,:,:].reshape(-1,1,84,84))
+
+    # Optimizaing the policy and value network
+    forward_mse = nn.MSELoss(reduction='none')
+    target_agent = Agent().to(device)
+    inds = np.arange(args.batch_size,)
+
+    rnd_next_obs = torch.FloatTensor(((b_obs.data.cpu().numpy()[:,3,:,:].reshape(-1, 1, 84, 84) - obs_rms.mean) / np.sqrt(obs_rms.var)).clip(-5, 5)).to(device)
+
+    for i_epoch_pi in range(args.update_epochs):
+        target_agent.load_state_dict(agent.state_dict())
+        np.random.shuffle(inds)
+        for start in range(0, args.batch_size, args.minibatch_size):
+            end = start + args.minibatch_size
+            minibatch_ind = inds[start:end]
+            predict_next_state_feature, target_next_state_feature = rnd_model(rnd_next_obs[minibatch_ind])
+            forward_loss = forward_mse(predict_next_state_feature, target_next_state_feature.detach()).mean(-1)
+            
+            mask = torch.rand(len(forward_loss)).to(device)
+            mask = (mask < args.update_proportion).type(torch.FloatTensor).to(device)
+            forward_loss = (forward_loss * mask).sum() / torch.max(mask.sum(), torch.Tensor([1]).to(device))
+            del mask
```

Comment:
I assume this will be taken care for by the garbage collector. I suggest removing it.